File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/settings/RecoveryCodeFragment.java
package io.muun.apollo.presentation.ui.settings;

import io.muun.apollo.R;
import io.muun.apollo.domain.errors.UserFacingError;
import io.muun.apollo.presentation.ui.settings.edit_password.BaseEditPasswordFragment;
import io.muun.apollo.presentation.ui.view.MuunButton;
import io.muun.apollo.presentation.ui.view.MuunRecoveryCodeBox;

import android.view.View;
import butterknife.BindView;
import butterknife.OnClick;

public class RecoveryCodeFragment extends BaseEditPasswordFragment<RecoveryCodePresenter>
        implements RecoveryCodeView {

    @BindView(R.id.enter_recovery_code_box)
    MuunRecoveryCodeBox recoveryCodeBox;

    @BindView(R.id.enter_recovery_code_continue)
    MuunButton continueButton;

    @Override
    protected void inject() {
        getComponent().inject(this);
    }

    @Override
    protected int getLayoutResource() {
        return R.layout.recovery_code_fragment;
    }

    @Override
    protected void initializeUi(View view) {
        super.initializeUi(view);

        recoveryCodeBox.setEditable(true);
        recoveryCodeBox.setOnEditedListener(this::onRecoveryCodeEdited);
        recoveryCodeBox.requestFocusOnFirstEditableSegment();
        recoveryCodeBox.setOnKeyboardNextListeners();
        recoveryCodeBox.setOnKeyboardDoneListener(() -> {
            if (continueButton.isEnabled()) {
                continueButton.callOnClick();
            }
        });
    }

    @Override
    public void onResume() {
        super.onResume();
        onRecoveryCodeEdited(recoveryCodeBox.getSegmentInputsContent()); // re-trigger validation
        recoveryCodeBox.requestFocusOnFirstEditableSegment();
    }

    @Override
    protected boolean blockScreenshots() {
        return true;
    }

    @OnClick(R.id.use_password)
    public void onUsePasswordButtonlick() {
        getParentActivity().onBackPressed();
    }

    @OnClick(R.id.enter_recovery_code_continue)
    public void onContinueButtonClick() {
        presenter.submitRecoveryCode(recoveryCodeBox.getSegmentInputsContent());
    }

    @Override
    public void setRecoveryCodeError(UserFacingError error) {
        recoveryCodeBox.setError(error);
    }

    @Override
    public void setConfirmEnabled(boolean isEnabled) {
        continueButton.setEnabled(isEnabled);
    }

    private void onRecoveryCodeEdited(String recoveryCodeString) {
        presenter.onRecoveryCodeEdited(recoveryCodeString);
    }
}

File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/settings/RecoveryCodeFragment.kt
package io.muun.apollo.presentation.ui.settings

import android.view.View
import butterknife.BindView
import io.muun.apollo.R
import io.muun.apollo.domain.errors.UserFacingError
import io.muun.apollo.presentation.ui.base.SingleFragment
import io.muun.apollo.presentation.ui.view.MuunButton
import io.muun.apollo.presentation.ui.view.MuunRecoveryCodeBox

class RecoveryCodeFragment : SingleFragment<RecoveryCodePresenter>(), RecoveryCodeView {

    @BindView(R.id.enter_recovery_code_box)
    lateinit var recoveryCodeBox: MuunRecoveryCodeBox

    @BindView(R.id.use_password)
    lateinit var usePasswordButton: MuunButton

    @BindView(R.id.enter_recovery_code_continue)
    lateinit var continueButton: MuunButton

    override fun inject() {
        component.inject(this)
    }

    override fun getLayoutResource(): Int =
        R.layout.recovery_code_fragment

    override fun initializeUi(view: View) {
        recoveryCodeBox.setEditable(true)
        recoveryCodeBox.setOnEditedListener { recoveryCodeString: String ->
            onRecoveryCodeEdited(recoveryCodeString)
        }
        recoveryCodeBox.requestFocusOnFirstEditableSegment()
        recoveryCodeBox.setOnKeyboardNextListeners()
        recoveryCodeBox.setOnKeyboardDoneListener {
            if (continueButton.isEnabled) {
                continueButton.callOnClick()
            }
        }

        usePasswordButton.setOnClickListener { onUsePasswordButtonClick() }
        continueButton.setOnClickListener { onContinueButtonClick() }
    }

    override fun onResume() {
        super.onResume()
        onRecoveryCodeEdited(recoveryCodeBox.segmentInputsContent) // re-trigger validation
        recoveryCodeBox.requestFocusOnFirstEditableSegment()
    }

    override fun blockScreenshots(): Boolean {
        return true
    }

    private fun onUsePasswordButtonClick() {
        parentActivity.onBackPressed()
    }

    private fun onContinueButtonClick() {
        presenter.submitRecoveryCode(recoveryCodeBox.segmentInputsContent)
    }

    override fun setRecoveryCodeError(error: UserFacingError?) {
        recoveryCodeBox.setError(error)
    }

    override fun setConfirmEnabled(isEnabled: Boolean) {
        continueButton.isEnabled = isEnabled
    }

    private fun onRecoveryCodeEdited(recoveryCodeString: String) {
        presenter.onRecoveryCodeEdited(recoveryCodeString)
    }
}